Add custom mock argument matching

jgraettinger opened the pull request:
Add an ArgumentMatcher interface and MatchedBy() helper to the mock
package, enabling custom fine-grained argument matching behaviors.

This is particularly helpful when building mocks over complex argument
types where deep equality is cumbersome or can't be used. Ex, a matcher
might match over just the the URL of an argument *http.Request.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/stretchr/testify/227)
<!-- Reviewable:end -->

ernesto-jimenez commented:
@jgraettinger I think the core idea is great, but I would make some changes.

Looking at the test, I think it would be nicer to do:

```go
mockedService.On("TheExampleMethod",
  MatchedBy(func(a int) bool {
    return a == 1
})
```

You can use the `reflect` package to look at the signature of the function sent to `MatchedBy` which means that we can check whether the type of the argument to the mock call matches the type of the argument of the function provided to `MatchedBy`. It also gets rid of the need for converting the `interface{}` to whatever type you are expecting.

I think it would make the feature nicer to use. What do you think?

ernesto-jimenez commented:
@jgraettinger I remembered `gocraft/web` uses the reflect package to check if a function matches certain signature.

You can see the code here: https://github.com/gocraft/web/blob/3fd2ffd35c9233bfe91caa84e9c09d7cf8fae964/router_setup.go#L318-L357

It should help.

Also, what use case do you have in mind for exporting the `ArgumentMatcher` interface?

jgraettinger commented:
@ernesto-jimenez Think your feedback here is great. I've implemented as described, and added a bit more testing here. Also opted to drop the ArgumentMatcher interface altogether. PTAL

ernesto-jimenez commented on mock/mock.go:
```diff
@@ -458,6 +458,47 @@ func AnythingOfType(t string) AnythingOfTypeArgument {
+		panic(fmt.Sprintf("assert: arguments: %s does not return a bool", fn))
+	}
+
+	return argumentMatcher{fn: reflect.ValueOf(fn)}
```
maybe we should store `reflect.Value(fn).Type().In(0)`?

jgraettinger replied:
Mind elaborating on the rationale here? As an optimization?

IANAE, but as I understand it once you have a reflect.Value, the heavy lifting of type introspection has already been done.

ernesto-jimenez replied:
nevermind, I missed one line where you were using more than `.Type().Int(0)`

ernesto-jimenez commented:
@jgraettinger looks good to me. The only thing missing is some documentation. Could you add some?

Thanks!
Ernesto
